### ISSUE TITLE
Slow counting.

### ISSUE THREAD

**Issue by ericrini** [rethinkdb/rethinkdb#3952]
Hi guys, looking for some help. Can anyone explain why this query doesn't seem to use an index and can take 10-20 seconds to run.

```r.db('ijoin')
  .table('plans')
   
  // One-To-Many
  .map(function (parent) {
    
    return parent.merge({
      investmentVehicles: r.db('ijoin')
          .table('investmentVehicles')
          .getAll(parent('id'), { index: 'planId' })
          .coerceTo('array')
    });
  })

  .count() ```

...but this one appears to work fine and easily counts 20,000+ records in 50ms?

```r.db('ijoin')
  .table('plans')
   
  // One-To-Many
  .map(function (parent) {
    
    return parent.merge({
      investmentVehicles: []
    });
  })

  .count() ```

**Comment by mlucy** [comment]
Hi @ericrini !

The first one is doing a range get for every document, while the second one is just doing a bit of processing (it probably isn't even loading the row off disk if your 20k records fit in cache).  50ms -> 10s is a factor of 200, which sounds about right to me for a sub-query that needs to go across the network vs. a sub-query that executes a small amount of code, especially if you're moving a lot of data around.  (I think the sub-query *is* using the index.  If you replaced the `get_all` with a `filter`, I would guess this query would get much, much slower.)

We could probably improve the performance of this case by either batching up the cross-network requests, but it's always going to be much slower than the second case.  (@danielmewes, do you think that performance improvement is worth doing right now?)

If you need to do this sort of query frequently, I'd recommend moving the `investmentVehicles` table into the `plans` table as a sub-document (or vice-versa depending on how your data is structured).

**Comment by danielmewes** [comment]
I wonder if evaluating the `map` in parallel would increase the throughput for this query (https://github.com/rethinkdb/rethinkdb/issues/2156).

Generally I think optimizing queries like this is worthwhile because the same optimizations will likely speed up all sorts of joins.
We should probably tackle this soon after 2.1.

**Comment by chrisvariety** [comment]
The map in your query is actually unnecessary -- is this any faster, or does it evaluate the same behind the scenes? We use this pattern all the time and haven't run into performance issues (yet)

```
r.db('ijoin')
  .table('plans')

  // One-To-Many
  .merge({
      investmentVehicles: r.db('ijoin')
          .table('investmentVehicles')
          .getAll(r.row('id'), { index: 'planId' })
          .coerceTo('array')
  })
```

Alternatively you can try the following, we do this a lot too, it's called 'sideloading', slightly different data structure but may be faster, worth a shot!

```
r.db('ijoin')
  .table('plans').coerceTo('array').do(function(plans) {
  return {
    plans: plans,
    investmentVehicles: r.db('ijoin').table('investmentVehicles').getAll(plans.concatMap(function(plan) {
  return plan('id');
}), {index: 'planId'}).coerceTo('array')
  };
})
```

Secondarily, I've found `.count()`s to be extremely slow compared to actually returning the results... worth a shot removing the count entirely.

**Comment by danielmewes** [comment]
> Secondarily, I've found `.count()`s to be extremely slow compared to actually returning the results... worth a shot removing the count entirely.

That might happen if not all results of the query are retrieved. In that case only the first few results of the query without `count` will be generated by the server, until more results are requested by the client. Could this be the case?
I would be very surprised if loading all results of a query was faster than running the same query with `count`.

**Comment by danielmewes** [comment]
I believe omitting the `map` and applying `merge` directly translates to the same query internally.

**Comment by mlucy** [comment]
@chrisvariety -- as you guessed, that `merge` is actually rewritten to a `.map.merge` internally.

**Comment by chrisvariety** [comment]
@danielmewes Ah yes, you're right, I take that back-- my specific scenario does involve pagination (which is itself very slow #3949 :wink:)

**Comment by ericrini** [comment]
I agree, it's a contrived example. Consider replacing count() with .skip(10,000) .limit(100). You get a similar level of slowness that grows linearly the larger the .skip() is.

**Comment by chrisvariety** [comment]
@ericrini All of a sudden we are in a pagination scenario :smile: You can see more info about `.skip()`'s linear performance including some workarounds at #3105

**Comment by PrakashThapa** [comment]
This  issue was never solved. I am very surprise. 
This query
```r.db("dbname").table("tablename").filter(function(f){return f("price").default(0).gt(40)})```
executed less than 8ms (server time). 
but once  ```count```  is connect like 
```r.db("dbname").table("tablename").filter(function(f){return f("price").default(0).gt(40)}).count()```
or 
```r.db("dbname").table("tablename").count(function(f){return f("price").default(0).gt(40)})```

It took more than 92ms why? can anyone give the proper reason? I am following rethinkdb since 1 and half year and same problem with ```count``` each time.

**Comment by mlucy** [comment]
@PrakashThapa -- RethinkDB streams data to the clients, so you'll get a cursor back as soon as the first batch of data is available in the `filter` case, but the `count` case doesn't return until the entire stream is consumed.  I would guess that if you wrote `r.db("dbname").table("tablename").filter(function(f){return f("price").default(0).gt(40)}).coerce_to('array')` to force the entire stream to be read before anything is returned to the client that the performance would be worse than the `count` performance.

**Comment by PrakashThapa** [comment]
@mlucy So there is no any alternative way to make count() faster? I want to just count the result that satisfied the condition. 

yes I tried with coerce_to('array') but it is still slower than count for server time. but better in round-trip time.

**Comment by mlucy** [comment]
@PrakashThapa -- There isn't a clever way to make that `count` query faster on your end, unfortunately.  If a lot of your rows are missing the `price` field then it will get faster when we fix https://github.com/rethinkdb/rethinkdb/issues/4271 .
